scripts/generate_matrix.py

```diff
@@ -0,0 +1,200 @@
+"""
+Script to generate a test matrix of Python and package versions for CI testing.
+
+This script fetches the list of supported Python versions and the versions of a
+specified package available on PyPI, determines compatible combinations based
+on package 'requires_python' metadata, and outputs a JSON matrix suitable for
+use in CI workflows.
+
+Features:
+- Generalized to work with any package (not just NumPy).
+- Caches HTTP requests to reduce network calls.
+- Prevents code duplication by using reusable functions.
+- Includes thorough doctests and documentation.
+
+Usage:
+    python generate_matrix.py [PACKAGE_NAME]
+
+Example:
+    python generate_matrix.py numpy
+
+"""
+
+import importlib.metadata
+import json
+import sys
+import urllib.error
+import urllib.request
+from functools import lru_cache
+import typing
+
+from packaging.version import Version, parse
+
+# Constants for caching
+CACHE_SIZE = 128  # Adjust as needed
+
+
+def get_minimum_python_version(package: str) -> Version:
+    """
+    Fetch the minimum Python version specified in pyproject.toml.
+
+    Returns:
+        str: The minimum required version of the package.
+
+    Examples:
+        >>> get_minimum_python_version()  # doctest: +ELLIPSIS
+        Version(major=3, ...)
+    """
+    raw_version = importlib.metadata.metadata(package)["Requires-Python"]
+    if "<" in raw_version:
+        raise NotImplementedError("Version specifier with upper bound not yet supported!")
+
+    return parse(raw_version.replace(">=", "").replace("~=", ""))
+
+
+def get_minimum_version(package: str, dependency: str) -> Version:
+    """
+    Fetch the minimum dependency version required by the given package.
+
+    Args:
+        package (str): The package name ('scipy' or package name).
+        dependency (str): The dependency name ('numpy' or dependency name).
+
+    Returns:
+        str: The minimum required version of the package.
+
+    Examples:
+        >>> get_minimum_version("scipy", "numpy")  # doctest: +ELLIPSIS
+        Version(major=1, ...)
+    """
+    np_minimum_dep = next(
+        req for req in importlib.metadata.requires(package) if req.startswith(dependency) and " extra " not in req
+    )
```

Comment:
in case you hate ruff for formatting it like this as well, you can work around it with a `# fmt: skip` pragma thingy, which applies to the entire statement, so e.g.

```suggestion
    np_minimum_dep = next(
        req
        for req in importlib.metadata.requires(package) 
        if req.startswith(dependency)
        and " extra " not in req
    )  # fmt: skip
```